test/utils/test-Utf8.cc

```diff
@@ -0,0 +1,222 @@
+
+            std::ifstream input_is{ //
+                std::string("test/fixtures/utf-8/") + file_entry.second,
+                std::ios_base::binary
+            };
+            REQUIRE(input_is);
+
+            const std::string str( //
+                (std::istreambuf_iterator<char>(input_is)),
+                std::istreambuf_iterator<char>());
+
+            tested b{ str.begin(), str.end() };
+            tested e{ str.end(), str.end() };
+
+            WHEN("the utf iterator to begin is incremented until it is equal to end")
+            {
+                int i = 0;
+                for (; b != e; ++b, ++i)
+                    ;
+
+                THEN("it has been incremented `wc -m $fixture` times")
+                {
+                    REQUIRE(file_entry.first == i);
+                }
+            }
+
+            WHEN("serialize(codepoint) is used on deref'd iterator")
+            {
+                THEN("the result is the same as the original input")
+                {
+                    auto p_exp = str.begin();
+                    for (; b != e; ++b) {
+                        std::vector<char> cs;
+
+                        // TODO @tjanc@ also test return value
+                        serialize(*b, std::back_insert_iterator<std::vector<char> >(cs));
+
+                        for (const char& c : cs) {
+                            if (c != *p_exp) // Catch REQUIRE not used; would clutter
+                                break;       // number of assertions in Catch output
+                            ++p_exp;
+                        }
+                    }
+
+                    // asserting we arrived at the end is sufficient
+                    REQUIRE(p_exp == str.end());
+                }
+            }
+        }
+    }
+}
+
+SCENARIO("iteration over non utf-8 strings", "[utf8]")
+{
+    using tested = input_iterator<std::string::const_iterator>;
+
+    GIVEN("a string ending prematurely; and an utf8 iterator to it")
+    {
+        std::string str{ "ab\xC2" };
+
+        tested b{ str.begin(), str.end() };
+
+        WHEN("the iterator is advanced two times")
+        {
+            using std::advance;
+            advance(b, 2);
+
+            THEN("it dereferences to the replacement character, 0xFFFD")
+            {
+                REQUIRE(0xFFFD == *b);
+            }
+
+            WHEN("incremented")
+            {
+                ++b;
+
+                THEN("it is equal to an utf8 iterator to the end of the string")
+                {
+                    tested end{ str.end(), str.end() };
+                    REQUIRE(b == end);
+                }
+            }
+        }
+    }
+
+    // TODO @tjanc@ test ending prematurely 3-byte
+    // TODO @tjanc@ test ending prematurely 4-byte
+    // TODO @tjanc@ test oversized 2-byte
+    // TODO @tjanc@ test oversized 3-byte
+    // TODO @tjanc@ test oversized 4-byte
+    // TODO @tjanc@ test surrogates as codepoints (3-byte)
```

Comment:
What are these about?

Reply:
Does the PR need to wait for them?

Reply:
This SHOULD be implemented to make the tests, although such a condition CANNOT occur as long as the parser implements utf-8 parsing correctly. Because that's not the case, I shall implement those tests. It's an oversight on my side, I should have grep'ed for `TODO`s in my source code.